lib/commands/context.js

```diff
@@ -0,0 +1,472 @@
+  }
+  await B.delay(timeout);
+  if (this.useNewSafari()) {
+    await this.typeAndNavToUrl();
+  } else if (parseInt(this.iosSdkVersion, 10) >= 7 && !this.caps.udid && this.caps.safari) {
+    await this.navToViewThroughFavorites();
+  } else {
+    await this.navToViewWithTitle(/.*/);
+  }
+};
+
+extensions.typeAndNavToUrl = async function () {
+  let initialUrl = this.caps.safariInitialUrl || `http://127.0.0.1:${this.caps.port}/welcome`;
+  let oldImpWait = this.implicitWaitMs;
+  this.implicitWaitMs = 7000;
+  let res = await this.findElement('name', 'URL');
+  this.implicitWaitMs = oldImpWait;
+  await this.nativeTap(res.value.ELEMENT);
+  res = await this.findElements('name', 'Address');
+  let addressEl = res.value[res.value.length -1].ELEMENT;
+  await this.setValueImmediate(addressEl, initialUrl);
+  res = await this.findElement('name', 'go');
+  await this.nativeTap(res.value.ELEMENT);
+  await this.navToViewWithTitle(/.*/i);
+  // Waits for page to finish loading.
+  await this.remote.pageUnload();
+};
+
+extensions.navToViewThroughFavorites = async function () {
+  logger.debug('We are on iOS7+ simulator: clicking apple button to get into a webview');
+  let oldImpWait = this.implicitWaitMs;
+  this.implicitWaitMs = 7000; // wait 7s for apple button to exist
+  try {
+    let res = await this.findElement('xpath', '//UIAScrollView[1]/UIAButton[1]');
+    if (res.status !== 0) {
+      throw new Error('Failed!');
+    }
+  } catch (err) {
+    let msg = 'Could not find button to click to get into webview. ' +
+              'Proceeding on the assumption we have a working one.';
+    logger.error(msg);
+    this.implicitWaitMs = oldImpWait;
+    return await this.navToViewWithTitle(/.*/i);
+  }
+  this.implicitWaitMs = oldImpWait;
+  try {
+    let res = await this.nativeTap(res.value.ELEMENT);
+    if (res.status !== 0) {
+      throw new Error('Failed!');
+    }
+  } catch (err) {
+    let msg = 'Could not click button to get into webview. ' +
+              'Proceeding on the assumption we have a working one.';
+    logger.error(msg);
+  }
+  await this.navToViewWithTitle(/apple/i);
+};
+
+extensions.navToViewWithTitle = async function (titleRegex) {
+  logger.debug('Navigating to most recently opened webview');
+  let start = Date.now();
+  let spinTime = 500;
+  let spinHandles = async () => {
+    let res;
+    try {
+      res = await this.getLatestWebviewContextForTitle(titleRegex);
+    } catch (err) {
+      throw new Error(`Could not navigate to webview! Err: ${err.message}`);
+    }
+    if (!res) {
+      if ((Date.now() - start) < 90000) {
+        logger.warn("Could not find any webviews yet, refreshing/retrying");
+        if (this.caps.udid || !this.caps.safari) {
+          await B.delay(spinTime);
+          return await spinHandles();
+        }
+        try {
+          res = await this.findUIElementOrElements('accessibility id', 'ReloadButton', '', false);
```

Comment:
surprised that this doesn't just return the value directly; i thought we'd moved away from internal functions knowing anything about jsonwp response structure